File: countingsort.py
import numpy as np
import time 

def countingSort(arr):
	max_element = int(max(arr))
	min_element = int(min(arr))
	range_of_elements = max_element - min_element + 1
	# Create a count array to store count of individual
	# elements and initialize count array as 0
	count_arr = [0 for _ in range(range_of_elements)]
	output_arr = [0 for _ in range(len(arr))]

	# Store count of each character
	for i in range(0, len(arr)):
		count_arr[arr[i]-min_element] += 1

	# Change count_arr[i] so that count_arr[i] now contains actual
	# position of this element in output array
	for i in range(1, len(count_arr)):
		count_arr[i] += count_arr[i-1]

	# Build the output character array
	for i in range(len(arr)-1, -1, -1):
		output_arr[count_arr[arr[i] - min_element] - 1] = arr[i]
		count_arr[arr[i] - min_element] -= 1

	# Copy the output array to arr, so that arr now
	# contains sorted characters
	for i in range(0, len(arr)):
		arr[i] = output_arr[i]

	return arr


str1 = ''

arr2 = [100,200,300,400,500,600,700,800,900,1000,2000,3000,4000,5000,6000,7000,8000,9000,10000,20000,30000,40000,50000,60000,70000,80000,90000,100000]

for canti_data in arr2:
	print(canti_data)
	arr = np.loadtxt(f"./data/archivo{canti_data}.txt", dtype=int)
	inicio = time.time()
	countingSort(arr)
	fin = time.time()
	result = f'{(fin - inicio)*1000}'
	if canti_data != 100000:
		str1+= str(canti_data) + ' ' +  str(result) + '\n'
	else:
	    str1+= str(result)

file = open("./ResCountingSortPython.txt", "w")
file. write(str1)    
file. close()

File: countingsort.cpp
#include <iostream>
#include <chrono>
#include <fstream>
#include <string>
#include <windows.h>
#include <limits.h>
#include <algorithm>
#include <vector>
using namespace std;
using namespace std::chrono;
float t0, t1;


void countingSort(vector<int>& arr)
{
    int max = *max_element(arr.begin(), arr.end());
    int min = *min_element(arr.begin(), arr.end());
    int range = max - min + 1;
 
    vector<int> count(range), output(arr.size());
    for (int i = 0; i < arr.size(); i++)
        count[arr[i] - min]++;
 
    for (int i = 1; i < count.size(); i++)
        count[i] += count[i - 1];
 
    for (int i = arr.size() - 1; i >= 0; i--) {
        output[count[arr[i] - min] - 1] = arr[i];
        count[arr[i] - min]--;
    }
	
    for (int i = 0; i < arr.size(); i++)
        arr[i] = output[i];

	
}

string getCurrentDir() {
    char buff[MAX_PATH];
    GetModuleFileName( NULL, buff, MAX_PATH );
    string::size_type position = string( buff ).find_last_of( "\\/" );
    return string( buff ).substr( 0, position);
}
int main()
{
	int arr2[] = {100,200,300,400,500,600,700,800,900,1000,2000,3000,4000,5000,6000,7000,8000,9000,10000,20000,30000,40000,50000,60000,70000,80000,90000,100000};
	int n = sizeof(arr2)/sizeof(arr2[0]);
	
	string str1 = "";

	for(int i = 0; i < n; i++){
		int num = arr2[i];
		// int num = arr2[0];
		vector<int> arr;
		ifstream file(getCurrentDir() + "\\data\\archivo"+to_string(num)+".txt");
		
		if(file.is_open())
		{
			for(int i = 0; i < num; ++i)
			{
				int valor = 0;

				file >> valor;

				arr.push_back(valor);
			}
			file.close();
		}
		cout << num << endl;

		auto t0= high_resolution_clock::now();

		countingSort(arr);
		
		auto t1= high_resolution_clock::now();
		
		auto ms_int = duration_cast<milliseconds>(t1 - t0);

		duration<double, std::milli> ms_double = t1 - t0;

		str1 = str1 +to_string(num) + " " + to_string(ms_double.count()) + "\n";
	}
	
	ofstream MyFile(getCurrentDir() + "\\ResCountingSortCPlusPlus.txt");
	MyFile << str1<< endl;
	MyFile.close();

	return 0;
}
